### PR TITLE
Tensor changes in master graph and ring buffer

### PR DESCRIPTION
**Pull request by fiona-gladwin**
- Change all instances of Image to rocalTensor

**In Master Graph**
- Removed the internal batch size related params and computations
- Removed the output routine for video and making use of single loader routine for image and vide pipeline
- Remove variables associated with the change
- Remove output tensor allocation and deallocation (used for output copy)
- Remove copy_data API
- Introduced get_output_tensors API to directly return output tensor list

**In Ring Buffer**
- Remove master host buffers
- Modified host sub buffers to be used similar to device sub buffers

### CODE REVIEW

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 1]
```diff
@@ -52,27 +52,17 @@ class MasterGraph
...
-                    float offset0, float offset1, float offset2, bool reverse_channels, RocalTensorDataType output_data_type);
-    Status copy_output(unsigned char* out_ptr, size_t out_size_in_bytes);
-    Status copy_out_tensor_planar(void *out_ptr, RocalTensorFormat format, float multiplier0, float multiplier1, float multiplier2,
-                    float offset0, float offset1, float offset2, bool reverse_channels, RocalTensorDataType output_data_type);
-    size_t output_width();
-    size_t output_height();
-    size_t output_byte_size();
-    size_t output_depth();
+    rocalTensorList *get_output_tensors();
+    std::vector<size_t> tensor_output_byte_size();
```
Comment (sampath1117) [inline rocAL/rocAL/include/pipeline/master_graph.h:56]: Can we change this to return the vector as a reference?
std::vector<size_t>& tensor_output_byte_size();

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp [hunk 2]
```diff
@@ -159,7 +151,7 @@ void RingBuffer::init(RocalMemType mem_type, void *devres, unsigned sub_buffer_s
 
         }
     }
-   else
+    else
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/ring_buffer.cpp:154]: Please remove the additional space

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp [hunk 3]
```diff
@@ -142,15 +134,15 @@ void RingBuffer::init(RocalMemType mem_type, void *devres, unsigned sub_buffer_s
         {
             cl_mem_flags flags = CL_MEM_READ_ONLY;
 
-            _dev_sub_buffer[buffIdx].resize(_sub_buffer_count);
-            for(unsigned sub_idx = 0; sub_idx < _sub_buffer_count; sub_idx++)
+            _dev_sub_buffer[buffIdx].resize(sub_buffer_count);
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/ring_buffer.cpp:137]: Please align the code under the for loop at L137

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp [hunk 4]
```diff
@@ -172,33 +164,30 @@ void RingBuffer::init(RocalMemType mem_type, void *devres, unsigned sub_buffer_s
 
         for(size_t buffIdx = 0; buffIdx < BUFF_DEPTH; buffIdx++)
         {
-            _dev_sub_buffer[buffIdx].resize(_sub_buffer_count);
-            for(unsigned sub_idx = 0; sub_idx < _sub_buffer_count; sub_idx++)
+            _dev_sub_buffer[buffIdx].resize(sub_buffer_count);
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/ring_buffer.cpp:167]: Please align the code under the for loop at L167

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp [hunk 5]
```diff
@@ -172,33 +164,30 @@ void RingBuffer::init(RocalMemType mem_type, void *devres, unsigned sub_buffer_s
...
+            for(unsigned sub_idx = 0; sub_idx < sub_buffer_count; sub_idx++)
             {
 
-                hipError_t err =  hipMalloc(&_dev_sub_buffer[buffIdx][sub_idx], sub_buffer_size);
-                //printf("allocated HIP device buffer <%d, %d, %d, %p>\n", buffIdx, sub_idx, sub_buffer_size, _dev_sub_buffer[buffIdx][sub_idx]);
+                hipError_t err =  hipMalloc(&_dev_sub_buffer[buffIdx][sub_idx], _sub_buffer_size[sub_idx]);
+                //printf("allocated HIP device buffer <%d, %d, %d, %p>\n", buffIdx, sub_idx, _sub_buffer_size[sub_idx], _dev_sub_buffer[buffIdx][sub_idx]);
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/ring_buffer.cpp:172]: Please remove the print statement if not needed

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp [hunk 6]
```diff
@@ -172,33 +164,30 @@ void RingBuffer::init(RocalMemType mem_type, void *devres, unsigned sub_buffer_s
...
                     _dev_sub_buffer.clear();
-                    THROW("hipMalloc of size " + TOSTR(sub_buffer_size) + " index " + TOSTR(sub_idx) +
+                    THROW("hipMalloc of size " + TOSTR(_sub_buffer_size[sub_idx]) + " index " + TOSTR(sub_idx) +
                           " failed " + TOSTR(err));
                 }
             }
         }
     }
-   else
+    else
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/ring_buffer.cpp:182]: Please remove the additional space

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 7]
```diff
@@ -277,20 +274,13 @@ MasterGraph::create_single_graph()
 MasterGraph::Status
 MasterGraph::build()
 {
-    if(_output_images.empty())
-        THROW("No output images are there, cannot create the pipeline")
+    if(_internal_tensor_list.empty())
+        THROW("No output tensors are there, cannot create the pipeline")
 
-    // Verify all output images have the same dimension, otherwise creating a unified tensor from them is not supported
-    _output_image_info = _output_images.front()->info();
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:284]: Is this check not needed in tensor pipeline?

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 8]
```diff
@@ -975,76 +480,65 @@ void MasterGraph::output_routine()
...
-                        _output_images[idx]->swap_handle(write_buffers[idx]);
-                    else
-                    {
-                        auto this_cycle_buffer_ptr = (unsigned char *) write_buffers[idx] + each_cycle_size * cycle_idx;
-                        _output_images[idx]->swap_handle(this_cycle_buffer_ptr);
-                    }
-                }
+            // Swap handles on the output tensor, so that new processed tensor will be written to the a new buffer
+            for (size_t idx = 0; idx < _internal_tensor_list.size(); idx++)
+            {
```
Comment (sampath1117) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:507]: Please remove the outer braces for this loop since it has only one line inside the loop

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 9]
```diff
@@ -315,47 +283,46 @@ template<> inline std::shared_ptr<Cifar10LoaderNode> MasterGraph::add_node(const
...
 
     return node;
 }
 
-#ifdef ROCAL_VIDEO
 /*
  * Explicit specialization for VideoLoaderNode
  */
-template<> inline std::shared_ptr<VideoLoaderNode> MasterGraph::add_node(const std::vector<Image*>& inputs, const std::vector<Image*>& outputs)
+template<> inline std::shared_ptr<VideoLoaderNode> MasterGraph::add_node(const std::vector<rocalTensor*>& inputs, const std::vector<rocalTensor*>& outputs)
```
Comment (swetha097) [inline rocAL/rocAL/include/pipeline/master_graph.h:294]: Please maintain a space after rocalTensor* like other places to maintain uniformity

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 10]
```diff
@@ -315,47 +283,46 @@ template<> inline std::shared_ptr<Cifar10LoaderNode> MasterGraph::add_node(const
...
-        THROW("A video loader already exists, cannot have more than one loader")
+    if(_loader_module)
+        THROW("A loader already exists, cannot have more than one loader")
 #if ENABLE_HIP || ENABLE_OPENCL
     auto node = std::make_shared<VideoLoaderNode>(outputs[0], (void *)_device.resources());
 #else    
     auto node = std::make_shared<VideoLoaderNode>(outputs[0], nullptr);
-#endif    
-    _video_loader_module = node->get_loader_module();
-    _video_loader_module->set_prefetch_queue_depth(_prefetch_queue_depth);
```
Comment (swetha097) [inline rocAL/rocAL/include/pipeline/master_graph.h:337]: Video changes were removed on purpose ?

File: rocAL/rocAL/include/pipeline/ring_buffer.h [hunk 11]
```diff
@@ -72,13 +72,12 @@ class RingBuffer
...
     bool full();
     const unsigned BUFF_DEPTH;
-    unsigned _sub_buffer_size;
+    std::vector<size_t> _sub_buffer_size;
     unsigned _sub_buffer_count;
     std::mutex _lock;
     std::condition_variable _wait_for_load;
     std::condition_variable _wait_for_unload;
     std::vector<std::vector<void*>> _dev_sub_buffer;
-    std::vector<void*> _host_master_buffers;
```
Comment (swetha097) [inline rocAL/rocAL/include/pipeline/ring_buffer.h:81]: Why have we removed this line ? The concept of master buffer has been removed ?

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 12]
```diff
@@ -130,8 +129,6 @@ MasterGraph::MasterGraph(size_t batch_size, RocalAffinity affinity, int gpu_id,
 #endif        
         _first_run(true),
         _processing(false),
-        _internal_batch_size(compute_optimum_internal_batch_size(batch_size, affinity)),
```
Comment (swetha097) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:133]: Do we not require to use batch_size anywhere in master_graph ?

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 13]
```diff
@@ -353,23 +350,16 @@ void MasterGraph::release()
...
+    _loader_module->shut_down();
     // release all openvx resources.
     vx_status status;
-    for(auto& image: _internal_images)
-        delete image;// It will call the vxReleaseImage internally in the destructor
-    for(auto& image: _output_images)
-        delete image;// It will call the vxReleaseImage internally in the destructor
-    deallocate_output_tensor();
-
+    for(auto& tensor: _internal_tensors)
```
Comment (swetha097) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:359]: We should release the meta-data as well right ?

Comment (swetha097) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:359]: Check if all the other resources that were allocated has been released or not

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 14]
```diff
@@ -495,24 +407,14 @@ MasterGraph::reset()
     if(_output_thread.joinable())
         _output_thread.join();
     _ring_buffer.reset();
+    _sequence_start_framenum_vec.clear();
```
Comment (swetha097) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:410]: If all the changes wrt to video was removed , why is this added / retained yet?

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 15]
```diff
@@ -535,428 +437,31 @@ MasterGraph::mem_type()
...
-    size_t size = output_byte_size();
-    if (out_size_in_bytes != (size *_output_images.size()))
-        return MasterGraph::Status::INVALID_ARGUMENTS;
-
-    _convert_time.start();
-
-#if ENABLE_OPENCL
-    if(processing_on_device_ocl())
+    std::vector<void*> output_ptr = _ring_buffer.get_read_buffers();
+    // TODO - check here if size of internal tensor and ring buffer is same?
```
Comment (swetha097) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:451]: Is this TODO comment handled ?

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 16]
```diff
@@ -975,76 +480,65 @@ void MasterGraph::output_routine()
...
-                _graph->process();
+                if (full_batch_meta_data)
+                    full_batch_meta_data->concatenate(_augmented_meta_data);
+                else
+                    full_batch_meta_data = _augmented_meta_data->clone();
             }
+            _process_time.start();
+            _graph->process();
+            _process_time.end();
+
```
Comment (swetha097) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:541]: Remove the extra line here

**Review by swetha097 — COMMENTED** [review]
Added Review comments

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 17]
```diff
@@ -975,76 +480,65 @@ void MasterGraph::output_routine()
...
-                if (!_processing)
-                    break;
+            if (!_processing)
+                break;
 
-                for(auto node: _nodes)
+            for(auto node: _nodes)
+            {
+                if(node->_is_ssd)
                 {
```
Comment (swetha097) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:517]: Braces can be removed for just one liner statements for "if" conditions

File: rocAL/rocAL/source/pipeline/master_graph.cpp [hunk 18]
```diff
@@ -975,76 +480,65 @@ void MasterGraph::output_routine()
...
-                update_node_parameters();
-                if(_augmented_meta_data)
+            update_node_parameters();
+            if(_augmented_meta_data)
+            {
+                if (_meta_data_graph)
                 {
-                    if (_meta_data_graph)
+                    if(_is_random_bbox_crop)
                     {
```
Comment (swetha097) [inline rocAL/rocAL/source/pipeline/master_graph.cpp:528]: Braces can be removed for just one liner statements for "if" conditions

**Review by swetha097 — COMMENTED** [review]
Added 2 more comments

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 19]
```diff
@@ -88,41 +78,27 @@ class MasterGraph
...
-    }
-    void set_output(Image* output_image);
+    void set_output(rocalTensor* output_tensor);
     bool empty() { return (remaining_count() < (_is_sequence_reader_output ? _sequence_batch_size : _user_batch_size)); }
-    size_t internal_batch_size() { return _internal_batch_size; }
     size_t sequence_batch_size() { return _sequence_batch_size; }
     std::shared_ptr<MetaDataGraph> meta_data_graph() { return _meta_data_graph; }
     std::shared_ptr<MetaDataReader> meta_data_reader() { return _meta_data_reader; }
     bool is_random_bbox_crop() {return _is_random_bbox_crop; }
-    void set_video_loader_flag() { _is_video_loader = true; }
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply rocAL/rocAL/include/pipeline/master_graph.h:103]: Yes

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp (hunk 2 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/ring_buffer.cpp:154]: Added additional space to align with the if condition

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp (hunk 5 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/ring_buffer.cpp:172]: The print statement in present in TOT as well hence I will retain it

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 7 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:284]: Yes removed intentionally

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 9 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/include/pipeline/master_graph.h:294]: It was not there previously so did not add that

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 10 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/include/pipeline/master_graph.h:337]: Ok Will mention in the description

File: rocAL/rocAL/include/pipeline/ring_buffer.h (hunk 11 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/include/pipeline/ring_buffer.h:81]: Yes the master buffers are removed from ring buffer

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 12 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:133]: internal batch size is not required in master graph

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 14 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:401]: Separate output routine for video and associated variables were removed. This is to be returned to user.

File: rocAL/rocAL/include/pipeline/master_graph.h (hunk 1 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/include/pipeline/master_graph.h:56]: Removed this API as it is not used anywhere

File: rocAL/rocAL/include/pipeline/master_graph.h [hunk 20]
```diff
@@ -131,15 +107,13 @@ class MasterGraph
...
     rocalTensorList _output_tensor_list;    //!< Keeps a list of ovx tensors(augmented outputs) that are to be passed to the user (there is an augmentation output batch per element in the list)
-    std::list<Image*> _internal_images;//!< Keeps all the ovx images (virtual/non-virtual) either intermediate images, or input images that feed the graph
+    std::list<rocalTensor*> _internal_tensors;  //!< Keeps all the ovx tensors (virtual/non-virtual) either intermediate tensors, or input tensors that feed the graph
     std::list<std::shared_ptr<Node>> _nodes;//!< List of all the nodes
     std::list<std::shared_ptr<Node>> _root_nodes;//!< List of all root nodes (image/video loaders)
     std::list<std::shared_ptr<Node>> _meta_data_nodes;//!< List of nodes where meta data has to be updated after augmentation
-    std::map<Image*, std::shared_ptr<Node>> _image_map;//!< key: image, value : Parent node
-    void * _output_tensor;//!< In the GPU processing case , is used to convert the U8 samples to float32 before they are being transfered back to host
+    std::map<rocalTensor*, std::shared_ptr<Node>> _tensor_map;//!< key: tensor, value : Parent node
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply rocAL/rocAL/include/pipeline/master_graph.h:116]: Comment has been added already

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp (hunk 4 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/ring_buffer.cpp:167]: There is not for loop before that

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp (hunk 3 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/ring_buffer.cpp:137]: I think it is aligned please lmk if it is wrong

File: rocAL/rocAL/source/pipeline/ring_buffer.cpp (hunk 6 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/ring_buffer.cpp:182]: Aligned with if condition

File: rocAL/rocAL/source/pipeline/master_graph.cpp (hunk 13 above)
Comment (fiona-gladwin) [reply rocAL/rocAL/source/pipeline/master_graph.cpp:350]: Metadata is not added in the PR

**Comment by fiona-gladwin** [comment]
Merged with Lakshmi's branch
